---
title: "deseq"
output:
  html_document: default
  pdf_document: default
  css: styles.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Single layer analysis 
### Deseq results 


```{r Enrichment,  results='hide', echo=FALSE, message=FALSE  }

data_dir='D:/DATADRIVE/Efi Athieniti/Documents/git/mofa/'
print(getwd())
source(paste0(data_dir, '/ppmi/deseq_analysis_setup.R'))


outdir_mirs<-paste0(outdir_orig, '/single/', 'mirnas_',VISIT, '_',MIN_COUNT_M, '_coh_',sel_coh_s, '_',des)
outdir_rnas<-paste0(outdir_orig, '/single/', 'rnas_',VISIT, '_',MIN_COUNT_G, '_coh_',sel_coh_s, '_',des)


#source('ppmi/RNAseq enrichment.R')

#
#p_enrich


```

```{r main, echo=FALSE}

ns<-table(se_filt$COHORT_DEFINITION)
ns<-paste(rownames(ns)[1], ns[1],', ',names(ns)[2],ns[2])
cat('Deseq settings: ', des , '\n', 
    'VISIT: ', VISIT, '\n', 
    ns)



```






#### PCA plots
##### Left: Filtering by significant features 

```{r plot, echo=FALSE, fig.width=10, fig.height=7, out.width = "40%"}

pca_pars<-paste0('_signif_', 2)
pca_plot<-paste0(pca_files, 'individuals', pca_pars, '.jpeg')
knitr::include_graphics(pca_plot, dpi = 96, )

pca_pars<-paste0('_signif_', 1)
pca_plot<-paste0(pca_files, 'individuals', pca_pars, '.jpeg')
knitr::include_graphics(pca_plot, dpi = 96, )



#show(pvol)




```

### Differential Expression 


```{r deseqtable, echo=FALSE, message=FALSE}
#install.packages('kableExtra')
suppressWarnings(library('kableExtra'))
library(dplyr)
library('DT')


log2fol_T_overall<-0.1
padj_T_overall<-.05

padj_T=padj_T_overall
log2fol_T=log2fol_T_overall
deseq_table<-paste0(outdir_s, '/results_df.csv')
deseq_res<-read.csv(deseq_table)


outdirs_all=list(outdir_rnas,outdir_mirs, outdir_s_p )
get_sig<-function(outdir_s){
    signif_file<-paste0(outdir_s,'/significant', padj_T, '_',log2fol_T, '.csv')
    deseq_table<-paste0(signif_file)
    ncols=10 
    deseq_res<-read.csv(deseq_table)[,1:ncols]
    
    padj_T_overall<-.01
    deseq_res_strict<-deseq_res[deseq_res$padj<padj_T_overall,]
    return(c(dim(deseq_res)[1], dim(deseq_res_strict)[1]))
   
  
}

table_sig<-sapply(outdirs_all, get_sig) 
cat( 'log2foldChange > ', log2fol_T_overall,  ', padj < ', padj_T_overall,'\n',
         'DE features: ', dim(deseq_res)[1])
  

#### bring the absolute high fol to the top 
#deseq_res_ord<-deseq_res %>%
#  arrange(., -abslog2pval)
colnames(table_sig)<-c('RNA', 'miRNAs', 'proteomics')
rownames(table_sig)<-c('0.05', '0.01')

table_sig %>% kable %>%
  kable_styling("striped", full_width = F) 


#datatable(deseq_res_ord, options = list(scrollY = "300px"), class = "display", fillContainer = T, 
       #width = NULL, height = 1000, editable = FALSE) 

#deseq_res_ord %>%
#  kable %>%
#  kable_styling("striped", full_width = F) %>% 
# scroll_box(width = "1000px", height = "300px")


#if (process_mirnas){COL='X'}else{COL='SYMBOL'}
  
#  top<-deseq_res_ord[,COL][1:10]
#  cat('Top 10 markers: ', top)




```




### Heatmaps 
```{r heatmaps, echo=FALSE, fig.show="hold", fig.width=6, fig.height=8, out.width = "50%"}

log2fol_T_hm<-0.1
padj_T_hm<-.05
order_by_metric='log2pval'
order_by_metric='padj_reverse'

cluster_cols=TRUE
filter_highly_var=FALSE
n_sig_f=30
most_var_t=FALSE

fname_hm<-paste0(outdir_rnas, '/heatmap3', '_',padj_T_hm,'_', log2fol_T_hm ,order_by_metric, 'high_var_' ,
              filter_highly_var,    '_', most_var_t, '_',  n_sig_f, cluster_cols, '.jpeg')

knitr::include_graphics(fname_hm, dpi = 96, )


fname_hm<-paste0(outdir_mirs, '/heatmap3', '_',padj_T_hm,'_', log2fol_T_hm ,order_by_metric, 'high_var_' ,
              filter_highly_var,    '_', most_var_t, '_',  n_sig_f, cluster_cols, '.jpeg')
knitr::include_graphics(fname_hm, dpi = 96, )

filter_highly_var=FALSE
n_sig_f=30
prefix='prot_';
fname_hm<-paste0(outdir_s_p, '/heatmap3', '_',padj_T_hm,'_', log2fol_T_hm ,order_by_metric, 'high_var_' ,
              filter_highly_var, '_', most_var_t, '_',  n_sig_f, cluster_cols, '.jpeg')

knitr::include_graphics(fname_hm, dpi = 96, )



#source(paste0(data_dir,'/ppmi/RNAseq enrichment.R'))
#show(dp)
#show(p_enrich)
#show(p2_tree)




```


```{r run_enrich, echo=FALSE, fig.show="hold", fig.width=6, fig.height=8, out.width = "50%"}




prefix='rnas_'
fname_vol<-paste0(outdir_rnas, '/EnhancedVolcano_edited_', prefix, VISIT,'.jpeg')
knitr::include_graphics(fname_vol, dpi = 96, )

prefix='mirnas_'
fname_vol<-paste0(outdir_mirs, '/EnhancedVolcano_edited_', prefix, VISIT,'.jpeg')
knitr::include_graphics(fname_vol, dpi = 96, )


prefix='prot_'
fname_vol<-paste0(outdir_s_p, '/EnhancedVolcano_edited_', prefix, VISIT,'.jpeg')
knitr::include_graphics(fname_vol, dpi = 96, )

#source(paste0(data_dir,'/ppmi/RNAseq enrichment.R'))
#show(dp)
#show(p_enrich)
#show(p2_tree)




```


## Enrichment analysis 
### Cluster Profiler, GSEA GO 

```{r run_enrich2, echo=FALSE, fig.width=8, fig.height=6,  out.width = "60%"}

padj_T=1;log2fol_T=0.00
order_by_metric<-'log2pval'

ONT='BP'
cat('Category:', ONT)

  outdir_enrich<-paste0(outdir_s,'/enrichment/')

if (process_mirnas){
  #padj_T=1; log2fol_T=0
  Category='GO Biological process (miRPathDB)'
  
  enrich_params<-paste0('_', padj_T, '_',  log2fol_T, '_',  order_by_metric)
  mir_results_file<-paste0(outdir_enrich, '/mirs_enrich_', enrich_params)
  
  mir_results_file_by_cat<-paste0(outdir_enrich, Category, '/mirs_enrich_', enrich_params)
  results_file<-mir_results_file_by_cat

  
}else{
  
    results_file<-paste0(outdir_enrich, '/gseGO', '_', ONT, '_', padj_T, '_',  log2fol_T, order_by_metric)

}


N=25
dp_file<-paste0(results_file, '_dot',  '.jpeg')
emap_file<-paste0(results_file, '_emap_', N,  '.jpeg')
N=30;N_RIDGE=25
p2_tree_file<-paste0(results_file, '_clusterplot_average_',N, '.jpeg')
r_p_file<-paste0(results_file, '_ridge_', N_RIDGE, '.jpeg')


knitr::include_graphics(dp_file, dpi = 96, )
knitr::include_graphics(emap_file, dpi = 96, )

knitr::include_graphics(p2_tree_file, dpi = 96, )
knitr::include_graphics(r_p_file, dpi = 96, )


```
